[no squash] Fix for blocks sometimes not being updated on the client plus some RemoteClient improvements

lhofhansl opened the pull request:
Real fix for #16008...

It turns out that #10537 was based on the assumption that `d` in `RemoteClient::getNextBlock(...)` is an euclidean distance. From code it is clear that it is not. Instead `d` is the distance to the next cube-surface (which is then filtered with isBlockInSight to only blocks in a sphere).

Before this PR changes made at some distance to a client would not be reflected for a while (something up to 30s).

Unfortunately this slows down _initial_ block loading, especially with large viewing_rang'es, somewhat (about 10-20%), as the server resets the nearest_unsent_d to closer values than before. I think that is not avoidable in order to be correct. (Still much faster than before #10537, though)
To regain the performance, this marks liquid and light updates as low_priority, for these kind of evens the server loop is not reset - with that initial loading is even faster then before.

This PR also include some improvements and simplifications to the code. 

- Goal of the PR

Fix visual glitches.

- How does the PR work?

See description above.

- Does it resolve any reported issue?
- Does this relate to a goal in [the roadmap](https://github.com/luanti-org/luanti/blob/master/doc/direction.md)?
- If not a bug fix, why is this PR needed? What usecases does it solve?

Related: #10537

## To do

This PR Ready for Review.

## How to test

This is easiest to test with two clients. Place player A to face in ordinal direction, let's say -Z. Place the player B into the lower-left of the visible cone of the player A, at some distance away. Now have player B mine a block. Notice that without this change it takes a while for the change to be reflected in player A's view, whereas with this PR it is near instance.

You can also test this with the ranged_weapons mod. Shot a rocket towards the a corner of your sight, and zoom into the location where the rocket would strike. Notice that without the PR the changed blocks will take a while to load.

lhofhansl commented on src/server/clientiface.cpp:
```diff
@@ -432,12 +432,17 @@ void RemoteClient::SentBlock(v3s16 p)
+		// will reset m_nearest_unsent_d to 0 anyway (set getNextBlocks).
+		p -= m_last_center;
+		m_nearest_unsent_d = std::min(m_nearest_unsent_d,
+				(s16)std::max(std::max(std::abs(p.X), std::abs(p.Y)), std::abs(p.Z)));
```
This is the main fix. Instead of the euclidean distance, we need to find the cube side from which we need restart the scanning. This is always <= `getDistanceFrom`. (If there's a nicer way to express this, I am all ears.)

lhofhansl commented on src/server/clientiface.cpp:
```diff
@@ -396,8 +381,6 @@ void RemoteClient::GetNextBlocks (
 	// emerging, continue next time browsing from here
 	if (nearest_emerged_d != -1) {
 		new_nearest_unsent_d = nearest_emerged_d;
-	} else if (nearest_emergefull_d != -1) {
```
This was not needed. Either nearest_emerged_d is set, then it is used and nearest_emergefull_d, or it is not set, in which case nearest_emergefull_d is the nearest (if set). The distinctions was pointless and made the code harder to understand.

lhofhansl commented on src/server/clientiface.cpp:
```diff
@@ -423,8 +406,7 @@ void RemoteClient::GetNextBlocks (
 {
-	if (m_blocks_sending.find(p) != m_blocks_sending.end()) {
-		m_blocks_sending.erase(p);
+	if (m_blocks_sending.erase(p) > 0) {
```
Avoiding two lookups into the sets.

lhofhansl commented on src/server/clientiface.cpp:
```diff
@@ -435,32 +418,33 @@ void RemoteClient::GotBlock(v3s16 p)
+	// and reset the scan loop if found
+	if (m_blocks_sending.erase(p) + m_blocks_sent.erase(p) > 0) {
+		m_nothing_to_send_pause_timer = 0;
+		if (!m_perf_tradeoffs) {
```
Server will get to the block in the next full round if performance tradeoffs are enabled.

lhofhansl commented:
@sfan5 Let me know what'd make review easier.

sfan5 commented:
I just need to have some time.

lhofhansl commented:
Heck I just added it. Have a look. I am not attached to every commit :)

sfan5 commented on src/server/clientiface.cpp:
```diff
@@ -62,7 +62,8 @@ RemoteClient::RemoteClient() :
 	m_max_gen_distance(g_settings->getS16("max_block_generate_distance")),
-	m_occ_cull(g_settings->getBool("server_side_occlusion_culling"))
+	m_occ_cull(g_settings->getBool("server_side_occlusion_culling")),
+	m_perf_tradeoffs(g_settings->getBool("performance_tradeoffs"))
```
nitpick: the description of "performance_tradeoffs" talks about visual glitches. maybe adjust it.

lhofhansl replied:
I think I am just going to remove this part. With the low_priority map edit events it's not needed anymore (or at least much less effective).

sfan5 commented on src/server/clientiface.cpp:
```diff
@@ -435,32 +418,33 @@ void RemoteClient::GotBlock(v3s16 p)
+			// will reset m_nearest_unsent_d to 0 anyway (set getNextBlocks).
+			p -= m_last_center;
+			m_nearest_unsent_d = std::min(m_nearest_unsent_d,
+					(s16)std::max(std::max(std::abs(p.X), std::abs(p.Y)), std::abs(p.Z)));
```
which distance metric is this? isn't `d` an euclidean distance?
surely this would be clearer, if so:
```cpp
s16 this_d = m_last_center.getDistanceFrom(p);
m_nearest_unsent_d = std::min(m_nearest_unsent_d, this_d);
```

sfan5 replied:
yeah, it was easy enough to miss for me.

sfan5 commented on src/server/clientiface.cpp:
```diff
@@ -435,32 +418,33 @@ void RemoteClient::GotBlock(v3s16 p)
+			// Reset to the maximum cube size that will
+			// guarantee that this block will the scanned again.
+			// Using m_last_center is OK, as a change in center
+			// will reset m_nearest_unsent_d to 0 anyway (set getNextBlocks).
```
```suggestion
			// will reset m_nearest_unsent_d to 0 anyway (see getNextBlocks).
```

sfan5 commented on src/server/clientiface.cpp:
```diff
@@ -435,32 +418,33 @@ void RemoteClient::GotBlock(v3s16 p)
+	// and reset the scan loop if found
+	if (m_blocks_sending.erase(p) + m_blocks_sent.erase(p) > 0) {
+		m_nothing_to_send_pause_timer = 0;
+		if (!m_perf_tradeoffs && !low_priority) {
```
What I think would be nice here is a short explanation on what happens if `low_priority == true`.
The send loop will start again but might only get to the updated block on the next iteration? Is that it?

lhofhansl commented on src/map.h:
```diff
@@ -47,6 +47,7 @@ struct MapEditEvent
 	MapNode n = CONTENT_AIR;
 	std::vector<v3s16> modified_blocks; // Represents a set
 	bool is_private_change = false;
+	bool low_priority = false;
```
Yep. Are you general OK with the idea?

sfan5 replied:
if it helps, sure.

lhofhansl commented:
Ok. Removed the `performance_tradeoffs` part, and added comment better explaining `low_priority` `MapEditEvent`s and explained why this is not using the euclidean distance.

sfan5 commented on src/server/clientiface.cpp:
```diff
@@ -435,32 +417,38 @@ void RemoteClient::GotBlock(v3s16 p)
+	// and reset the scan loop if found
+	if (m_blocks_sending.erase(p) + m_blocks_sent.erase(p) > 0) {
+		m_nothing_to_send_pause_timer = 0;
+		// If this is a low_pririty event, do not reset m_nearest_unsent_d.
```
```suggestion
		// If this is a low priority event, do not reset m_nearest_unsent_d.
```

sfan5 commented on src/server/clientiface.cpp:
```diff
@@ -435,32 +417,38 @@ void RemoteClient::GotBlock(v3s16 p)
+			// Using m_last_center is OK, as a change in center
+			// will reset m_nearest_unsent_d to 0 anyway (see getNextBlocks).
+			p -= m_last_center;
+			s16 this_d = (s16)std::max(std::max(std::abs(p.X), std::abs(p.Y)), std::abs(p.Z));
```
```suggestion
			s16 this_d = std::max({std::abs(p.X), std::abs(p.Y), std::abs(p.Z)});
```

lhofhansl replied:
I suppose this is slight more expensive as it has construct an array first... But I think that's fine, and it's more readable.

sfan5 commented:
Do you plan the squash the commits in some way?

lhofhansl commented:
Happy to squash. Or maybe I can squash to three commits: (1) the various random fixes, (2) the non-euclidean distance, (3) the low_priority changes.

sfan5 commented:
Sure. Just didn't want "add some comments" to stick around as separate.
<hr>
With this PR applied when you just join a world you stare into the void for a few seconds, because the code has <i>just</i> started the pause timer:

![screenshot_20250416_204648](https://github.com/user-attachments/assets/e4862433-6160-4b61-942e-f69f891c4f3a)
how can we avoid this?

debug patch:
```diff
@@ -90,6 +90,7 @@ void RemoteClient::GetNextBlocks (
                std::vector<PrioritySortedBlockTransfer> &dest)
 {
        // Increment timers
+       bool xx = m_nothing_to_send_pause_timer > 0;
        m_nothing_to_send_pause_timer -= dtime;
        m_map_send_completion_timer += dtime;
 
@@ -105,6 +106,8 @@ void RemoteClient::GetNextBlocks (
 
        if (m_nothing_to_send_pause_timer >= 0)
                return;
+       if (xx)
+               rawstream << "pause end\n";
 
        RemotePlayer *player = env->getPlayer(peer_id);
        // This can happen sometimes; clients and players are not in perfect sync.
@@ -385,6 +388,7 @@ void RemoteClient::GetNextBlocks (
                if (d > full_d_max) {
                        new_nearest_unsent_d = 0;
                        m_nothing_to_send_pause_timer = 2.0f;
+                       rawstream << "pausing send for 2s\n";
                        infostream << "Server: Player " << m_name << ", peer_id=" << peer_id
                                << ": full map send completed after " << m_map_send_completion_timer
                                << "s, restarting" << std::endl;
```

lhofhansl commented:
> With this PR applied when you just join a world you stare into the void for a few seconds, because the code has just started the pause timer: 

Why would this PR change this?
Oh.. Because (1) the emerge thread loads a blocks, triggers the dispatch event, (2) in SetBlockNotSent the block has not yet been by client (so it's not in the sending/sent list), (3) hence the m_nothing_to_send_pause_timer is not reset.
Hmm... Lemme think on that. I think the best way is to defensive move the `m_nothing_to_send_pause_timer = 0` outside of the if-statement.

lhofhansl commented:
Re: Squashing. I messed up the commits already. The comment commit has a logic change too.

lhofhansl commented:
Ok... Fixed and squashed into two commits. The (1) random fixes and (2) the non-euclidean stuff paired with the low-priority perf improvements.

lhofhansl commented:
How is this looking @sfan5 . Should fix your delayed initial visibility issue.

lhofhansl commented:
Have another looksee @sfan5 ? I think this is an important fix, especially for multi-user. If possible I'd want it it in 5.12.

lhofhansl commented:
Ok. Properly separated into three commits now:
1. RemoteClient improvements
2. Fox the nearest_unsent_d calculation
3. Mark light and liquid related updates as low_priority

Plus: Rebased

sfan5 commented:
> I think the best way is to defensive move the m_nothing_to_send_pause_timer = 0 outside of the if-statement.

sure, that works. Maybe a good idea to leave a comment for someone to investigate later?

lhofhansl commented:
> sure, that works. Maybe a good idea to leave a comment for someone to investigate later?

I looked into this a bit more ... Right now this is working as intended: The emerge thread signals to the RemoteClient that there are new blocks need to be sent to the client by calling `SetBlock(s)NotSent`.

Right now every client is notified of every block from every other client - even if far away or otherwise not visible. This resets the scan loop, but causes no other work.
That part I wanted to do only if actually had been sent to the client. That works in almost all cases, except for the initial load. I'll continue to think about this.
